parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/CleanUtil.java

```diff
@@ -26,45 +26,84 @@
+    Method cleanMethodLocal = null;
+    Object unsafeLocal = null;
+    Method invokeCleanerMethodLocal = null;
+    if (majorVersion >= 9) {
+      try {
+        final Class<?> clazz = Class.forName("sun.misc.Unsafe");
+        final Field theUnsafe = clazz.getDeclaredField("theUnsafe");
+        theUnsafe.setAccessible(true);
+        unsafeLocal = theUnsafe.get(null);
+        invokeCleanerMethodLocal = clazz.getMethod("invokeCleaner", ByteBuffer.class);
+        invokeCleanerMethodLocal.invoke(unsafeLocal, tempBuffer);
+      } catch (IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException | NoSuchMethodException | SecurityException
+        | ClassNotFoundException | NoSuchFieldException e) {
+        logger.warn("Cannot use direct ByteBuffer cleaner, memory leaking may occur", e);
+        unsafeLocal = null;
+        invokeCleanerMethodLocal = null;
+      }
+    } else {
+      try {
+        cleanerMethodLocal = tempBuffer.getClass().getMethod("cleaner");
+        cleanerMethodLocal.setAccessible(true);
+        final Object cleanerObject = cleanerMethodLocal.invoke(tempBuffer);
+        cleanMethodLocal = cleanerObject.getClass().getMethod("clean");
+        cleanMethodLocal.invoke(cleanerObject);
+      } catch (NoSuchMethodException | SecurityException | IllegalAccessException |
+        IllegalArgumentException | InvocationTargetException e) {
+        logger.warn("Cannot use direct ByteBuffer cleaner, memory leaking may occur", e);
+        cleanerMethodLocal = null;
+        cleanMethodLocal = null;
+      }
     }
-    CLEANER_FIELD = cleanerField;
-    CLEAN_METHOD = cleanMethod;
+    cleanerMethod = cleanerMethodLocal;
+    cleanMethod = cleanMethodLocal;
+    unsafe = unsafeLocal;
+    invokeCleanerMethod = invokeCleanerMethodLocal;
   }
 
-  public static void clean(ByteBuffer buffer) {
-    if (CLEANER_FIELD == null || CLEAN_METHOD == null) {
-      return;
-    }
-    try {
-      Object cleaner = CLEANER_FIELD.get(buffer);
-      CLEAN_METHOD.invoke(cleaner);
-    } catch (IllegalAccessException | InvocationTargetException | NullPointerException e) {
-      // Ignore clean failure
-      logger.warn("Clean failed for buffer " + buffer.getClass().getSimpleName(), e);
+  private CleanUtil() {
+    // Hide the default constructor since this is a utility class.
+  }
+
+  public static void cleanDirectBuffer(ByteBuffer buf) {
+    if (cleanMethod != null) {
+      try {
+        cleanMethod.invoke(cleanerMethod.invoke(buf));
+      } catch (IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException | SecurityException e) {
+        // Ignore
+      }
+    } else if (invokeCleanerMethod != null) {
+      try {
+        invokeCleanerMethod.invoke(unsafe, buf);
+      } catch (IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException | SecurityException e) {
+        // Ignore
+        e.printStackTrace();
```

Comment:
I would not print anything to stderr. I would suggest using a logger and use it in both cases (`cleanMethod` or `invokeCleanerMethod`)

Reply:
Good point, I've updated the code to write it to the logger.